File: sdk/go/internal/tx/messages.go
package tx

// MessageEncoder encodes a specific message type into protobuf bytes.
type MessageEncoder interface {
	// TypeURL returns the protobuf type URL (e.g., "/seocheon.activity.v1.MsgSubmitActivity").
	TypeURL() string
	// Encode returns the protobuf-encoded message bytes.
	Encode() []byte
}

// MsgSubmitActivity encodes /seocheon.activity.v1.MsgSubmitActivity.
// Fields: submitter(1, string), activity_hash(2, string), content_uri(3, string)
type MsgSubmitActivity struct {
	Submitter    string
	ActivityHash string
	ContentURI   string
}

func (m *MsgSubmitActivity) TypeURL() string {
	return "/seocheon.activity.v1.MsgSubmitActivity"
}

func (m *MsgSubmitActivity) Encode() []byte {
	return ConcatBytes(
		EncodeFieldString(1, m.Submitter),
		EncodeFieldString(2, m.ActivityHash),
		EncodeFieldString(3, m.ContentURI),
	)
}

// MsgWithdrawNodeCommission encodes /seocheon.node.v1.MsgWithdrawNodeCommission.
// Fields: operator(1, string)
type MsgWithdrawNodeCommission struct {
	Operator string
}

func (m *MsgWithdrawNodeCommission) TypeURL() string {
	return "/seocheon.node.v1.MsgWithdrawNodeCommission"
}

func (m *MsgWithdrawNodeCommission) Encode() []byte {
	return ConcatBytes(
		EncodeFieldString(1, m.Operator),
	)
}

// Coin represents a cosmos.base.v1beta1.Coin.
// Fields: denom(1, string), amount(2, string)
type Coin struct {
	Denom  string
	Amount string
}

func (c *Coin) Encode() []byte {
	return ConcatBytes(
		EncodeFieldString(1, c.Denom),
		EncodeFieldString(2, c.Amount),
	)
}

// MsgSend encodes /cosmos.bank.v1beta1.MsgSend.
// Fields: from_address(1, string), to_address(2, string), amount(3, repeated Coin)
type MsgSend struct {
	FromAddress string
	ToAddress   string
	Amount      []Coin
}

func (m *MsgSend) TypeURL() string {
	return "/cosmos.bank.v1beta1.MsgSend"
}

func (m *MsgSend) Encode() []byte {
	parts := [][]byte{
		EncodeFieldString(1, m.FromAddress),
		EncodeFieldString(2, m.ToAddress),
	}
	for _, coin := range m.Amount {
		coinBytes := coin.Encode()
		parts = append(parts, EncodeFieldBytes(3, coinBytes))
	}
	return ConcatBytes(parts...)
}

File: sdk/csharp/src/Internal/Tx/Messages.cs
namespace Seocheon.Sdk.Internal.Tx;

/// <summary>
/// Protobuf message encoders for Seocheon transaction types.
/// </summary>
public static class Messages
{
    // === Message Type URLs ===

    public const string TypeMsgSubmitActivity = "/seocheon.activity.v1.MsgSubmitActivity";
    public const string TypeMsgWithdrawRewards = "/seocheon.node.v1.MsgWithdrawNodeCommission";
    public const string TypeMsgConfirmDelegation = "/seocheon.node.v1.MsgConfirmDelegation";
    public const string TypeMsgSend = "/cosmos.bank.v1beta1.MsgSend";

    /// <summary>
    /// Encodes MsgSubmitActivity.
    /// Fields: 1=submitter, 2=activity_hash, 3=content_uri
    /// </summary>
    public static byte[] EncodeMsgSubmitActivity(string submitter, string activityHash, string contentUri)
    {
        return Protobuf.Concat(
            Protobuf.EncodeString(1, submitter),
            Protobuf.EncodeString(2, activityHash),
            Protobuf.EncodeString(3, contentUri)
        );
    }

    /// <summary>
    /// Encodes MsgWithdrawNodeCommission.
    /// Fields: 1=operator_address
    /// </summary>
    public static byte[] EncodeMsgWithdrawRewards(string operatorAddress)
    {
        return Protobuf.EncodeString(1, operatorAddress);
    }

    /// <summary>
    /// Encodes MsgConfirmDelegation.
    /// Fields: 1=delegator_address, 2=validator_address
    /// </summary>
    public static byte[] EncodeMsgConfirmDelegation(string delegatorAddress, string validatorAddress)
    {
        return Protobuf.Concat(
            Protobuf.EncodeString(1, delegatorAddress),
            Protobuf.EncodeString(2, validatorAddress)
        );
    }

    /// <summary>
    /// Encodes MsgSend.
    /// Fields: 1=from_address, 2=to_address, 3=amount (repeated Coin)
    /// Coin: 1=denom, 2=amount
    /// </summary>
    public static byte[] EncodeMsgSend(string fromAddress, string toAddress, string amount, string denom)
    {
        var coin = Protobuf.Concat(
            Protobuf.EncodeString(1, denom),
            Protobuf.EncodeString(2, amount)
        );

        return Protobuf.Concat(
            Protobuf.EncodeString(1, fromAddress),
            Protobuf.EncodeString(2, toAddress),
            Protobuf.EncodeMessage(3, coin)
        );
    }
}
